source/extensions/filters/udp/dns_filter/dns_filter.cc

```diff
@@ -76,21 +109,149 @@ void DnsFilter::onData(Network::UdpRecvData& client_request) {
-  // TODO(abaptiste): serialize and return a response to the client
 
+  // Serializes the generated response to the parsed query from the client. If there is a
+  // parsing error or the incoming query is invalid, we will still generate a valid DNS response
+  message_parser_.buildResponseBuffer(query_context, response);
   Network::UdpSendData response_data{query_context->local_->ip(), *(query_context->peer_),
                                      response};
   listener_.send(response_data);
 }
 
+DnsLookupResponseCode DnsFilter::getResponseForQuery(DnsQueryContextPtr& context) {
+  /* It appears to be a rare case where we would have more than one query in a single request.
+   * It is allowed by the protocol but not widely supported:
+   *
+   * See: https://www.ietf.org/rfc/rfc1035.txt
+   *
+   * The question section is used to carry the "question" in most queries,
+   * i.e., the parameters that define what is being asked. The section
+   * contains QDCOUNT (usually 1) entries.
+   */
+  for (const auto& query : context->queries_) {
+    // Try to resolve the query locally. If forwarding the query externally is disabled we will
+    // always attempt to resolve with the configured domains
+    if (isKnownDomain(query->name_) || !config_->forwardQueries()) {
+      // Determine whether the name is a cluster. Move on to the next query if successful
+      if (resolveViaClusters(context, *query)) {
+        continue;
+      }
+
+      // Determine whether we an answer this query with the static configuration
+      if (resolveViaConfiguredHosts(context, *query)) {
+        continue;
+      }
+    }
+    // TODO(abaptiste): resolve the query externally
+  }
+
+  if (context->answers_.empty()) {
+    return DnsLookupResponseCode::Failure;
+  }
+  return DnsLookupResponseCode::Success;
+}
+
+uint32_t DnsFilter::getDomainTTL(const absl::string_view domain) {
+  const auto& domain_ttl_config = config_->domainTtl();
+  const auto& iter = domain_ttl_config.find(domain);
+
+  uint32_t ttl;
+  if (iter == domain_ttl_config.end()) {
+    ttl = static_cast<uint32_t>(DEFAULT_RESOLVER_TTL.count());
+  } else {
+    ttl = static_cast<uint32_t>(iter->second.count());
+  }
+
+  return ttl;
+}
+
+bool DnsFilter::isKnownDomain(const absl::string_view domain_name) {
+  const auto& known_suffixes = config_->knownSuffixes();
+
+  // If we don't have a list of whitelisted domain suffixes, we will resolve the name with an
+  // external DNS server
+  if (known_suffixes.empty()) {
+    ENVOY_LOG(debug, "Known domains list is empty");
+    return false;
+  }
+
+  // TODO(abaptiste): Use a trie to find a match instead of iterating through the list
+  for (auto& suffix : known_suffixes) {
+    if (suffix->match(domain_name)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool DnsFilter::resolveViaClusters(DnsQueryContextPtr& context, const DnsQueryRecord& query) {
+  Upstream::ThreadLocalCluster* cluster = cluster_manager_.get(query.name_);
```

Comment:
Expecting the cluster name to be the same as the dns query name is not reasonable. I might want to map multiple dns entries to a single cluster.

The DnsEndpoint has endpoint_config as a one_of. The AddressList is the only field today. A new string with the cluster name must be the other way to map this I think. 

Any thoughts?
https://github.com/envoyproxy/envoy/pull/11074/files?file-filters%5B%5D=.cc&file-filters%5B%5D=.proto&file-filters%5B%5D=No+extension#diff-df2f505fefc41f4bafa39601bef65f14R51

Reply:
OK, I see what you mean.  I would have to work on that.  Would you settle for a TODO to add "ClusterEndpoint" or something along those lines to extend the endpoint config?

Reply:
Isn't the change minor to include a string for cluster_name as a oneof field? Would be good to get that in with this change

Reply:
I would not call the change minor.  I will work on it, but I certainly prefer to do this separately from this PR.

Reply:
Sure, that's fine

Reply:
I'll include this change in the next PR commit.